### ISSUE TITLE
How to implement customized loss function?

### ISSUE THREAD

**Issue by atique81** [vlfeat/matconvnet#164]
Hi all,
I am very new to matconvnet and still getting familiar with. I am trying to implement a customized loss function which is the squared hinge-loss function, also known as the L2-SVM loss as follows.

function [loss, dzdx] = squaredHingeLoss(X, labels, dzdy)
a = squeeze(X);
a = max(0, 1-a.*labels)^2;
loss = sum(a); % summing over all the images in a batch
if nargin<=2 retrun; end

dzdx = 2a(-labels);
dzdx = reshape(dzdx, size(X));

My question is, do I need to divide the dzdx by the batch size like this  dzdx = dzdx/size(X, 4);

**Comment by lenck** [comment]
Hi, if you are using the `cnn_train` provided in the examples, then you don't have to, as the division by the batch size is done there when updating the weights.

But in general, this is a constant during training, so only what you achieve is invariance to the batch size (which is quite useful, however :) ).

**Comment by atique81** [comment]
Hi @lenck ,
Thanks for reliving the issue. However, I think that dividing the derivatives by batchSize only works for coarse predictions, and not dense predictions (e.g.; computing loss and derivative for every pixel of an image). Therefore, instead of dividing it by the batchSize, I think it should be divided by the number of elements in the derivatives.

**Comment by lenck** [comment]
Yes, you are right about that. Without that it may break as you basically have different effective batch size if you pass differently sized images. But in some cases, e.g. in `vl_nnpdist` you may not want to aggregate the loss as you want to pass it further. So it really depends on the application, in the end...

But maybe, this question would be more appropriate on the discussion forum, as it is not really an issue...

**Comment by atique81** [comment]
Yup, I agree. Thanks.

**Comment by lenck** [comment]
I hope it is OK if I will close it... :)

**Comment by atique81** [comment]
yup, no problem. Go ahead.
